tommath.h

```diff
@@ -269,38 +269,60 @@ mp_bool mp_isodd(const mp_int *a) MP_WUR;
 /* set to zero */
 void mp_zero(mp_int *a);
 
-/* set to a digit */
-void mp_set(mp_int *a, mp_digit b);
-
-/* set a double */
-mp_err mp_set_double(mp_int *a, double b) MP_WUR;
-
-/* set a 32-bit const */
-/* TODO void - never fails */ mp_err mp_set_int(mp_int *a, unsigned long b);
-
-/* set a platform dependent unsigned long value */
-/* TODO void - never fails */ mp_err mp_set_long(mp_int *a, unsigned long b);
-
-/* set a platform dependent unsigned long long value */
-/* TODO void - never fails */ mp_err mp_set_long_long(mp_int *a, unsigned long long b);
-
-/* get a double */
+/* get and set doubles */
 double mp_get_double(const mp_int *a) MP_WUR;
+mp_err mp_set_double(mp_int *a, double b) MP_WUR;
 
-/* get a 32-bit value */
-unsigned long mp_get_int(const mp_int *a) MP_WUR;
-
-/* get a platform dependent unsigned long value */
-unsigned long mp_get_long(const mp_int *a) MP_WUR;
-
-/* get a platform dependent unsigned long long value */
-unsigned long long mp_get_long_long(const mp_int *a) MP_WUR;
-
-/* initialize and set a digit */
-mp_err mp_init_set(mp_int *a, mp_digit b) MP_WUR;
-
-/* initialize and set 32-bit value */
-mp_err mp_init_set_int(mp_int *a, unsigned long b) MP_WUR;
+/* get integer, set integer and init with integer (int32_t) */
+int32_t mp_get_i32(const mp_int *a) MP_WUR;
+void mp_set_i32(mp_int *a, int32_t b);
+mp_err mp_init_i32(mp_int *a, int32_t b) MP_WUR;
+
+/* get integer, set integer and init with integer (uint32_t) */
+uint32_t mp_get_u32(const mp_int *a) MP_WUR;
+void mp_set_u32(mp_int *a, uint32_t b);
+mp_err mp_init_u32(mp_int *a, uint32_t b) MP_WUR;
+
+/* get integer, set integer and init with integer (int64_t) */
+int64_t mp_get_i64(const mp_int *a) MP_WUR;
+void mp_set_i64(mp_int *a, int64_t b);
+mp_err mp_init_i64(mp_int *a, int64_t b) MP_WUR;
+
+/* get integer, set integer and init with integer (uint64_t) */
+uint64_t mp_get_u64(const mp_int *a) MP_WUR;
+void mp_set_u64(mp_int *a, uint64_t b);
+mp_err mp_init_u64(mp_int *a, uint64_t b) MP_WUR;
+
+/* get integer, set integer and init with integer (int) */
+#define mp_get_i(a)        ((int)mp_get_i32(a))
```

Comment:
Well this is some kind of philosophical question. These days I mostly write code with precise types int32_t. The only imprecise types I am using are intptr_t, size_t etc. This way I make it very clear if a value fits inside a variable. However I am also usually write code only for 32 and 64 bit architectures and want to stay portable across those. Usually they are also int=int32_t always. Performance wise int32 vs int64 does not make a difference on the 64 bit archs. I only have to use int64 when I really need it then.

If I want to code against architectures with int=int16_t or stay portable across 16/32/64 bit, one has to probably use imprecise types like int and long in order to use the most efficient int type for loops etc. However one takes the overflow risk then.

And there are probably many people who don't want the precise types. For those people I added this API, but it is also somehow low cost, since it is just a macro which selects the underlying function or in the i32 case, just an alias. Also note that tommath takes the imprecise stand as of now. We are not using precise types internally for example. This becomes a hypothetical problem with numbers larger than 2^(16*65535) or so ;)

Reply:
In other words I've read the following statements.

1. "Use stdint.h types to stay portable"
2. "Use int/long types to stay portable"

:see_no_evil: